keps/sig-windows/2802-identify-windows-pods-apiserver-admission/README.md

```diff
@@ -0,0 +1,344 @@
+This can be mitigated by 
+- Having appropriate warnings in the releases before we beta and GA will mitigate the problem.
+- Having an example out-of-tree admission webhook which can detect such workloads and update them.(Having an in-tree admission plugin which mutates the pod spec to have appropriate Windows specific RuntimeClass may be too hard to do generically and not transparent to the user)
+
+
+## Design Details
+
+
+We can piggyback on the existing RuntimeClass admission controller to query for the RuntimeClass and see if it has
+`kubernetes.io/os: windows` in it's scheduling field. 
+TODO: We need to get an ack from sig-auth to use it.
+
+If we cannot piggyback on the RuntimeClass controller we need to build our own validating admission controller for Windows specific pods.
+
+### Test Plan
+
+
+Whatever path we choose, unit and e2e tests are a hard requirement to ensure smooth
+transition to make RuntimeClasses default choice for the users.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial e2e tests completed and enabled
+- Users getting a warning to switch to RuntimeClasses in beta release
+
+#### Alpha -> Beta Graduation
+- Gather feedback from end users
+- Tests are in Testgrid and linked in the KEP
+- Deny admission to pods that have only `kubernetes.io/os: windows` in nodeSelector
+  but don't have an associated RuntimeClass with `kubernetes.io/os: windows` in
+  scheduling field.
+
+#### Alpha -> Beta Graduation
+- 2 examples of end users using this field
+
+### Upgrade / Downgrade Strategy
+
+- Upgrades:
+  When upgrading from a release without this feature, to a release with `RuntimeClassesForWindows` enabled, we will honor both nodeSelector and RuntimeClasses
+  but a warning will be sent to the user that `honoring explicit nodeSelector without Runtimeclasses will be stopped from next release` if the pod has just nodeSelector set. This ensures users have enough time to move without breaking their workloads.
+- Downgrades:
+  When downgrading from a release with this feature to a release without `RuntimeClassesForWindows`, the existing behavior will continue where both nodeSelector and RuntimeClasses are honoured without a warning message.
+
+### Version Skew Strategy
+
+If the feature gate is enabled, only admission controller(either RuntimeClass or Windows specific admission controller) should be impacted. This feature may have some kubelet implications as the code to strip security constraints based on OS can be removed.
+
+
+## Production Readiness Review Questionnaire
+
+
+### Feature Enablement and Rollback
+
+
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: RuntimeClassesForWindows
```

Comment:
I would rather we go with a different name. `RuntimeClassesForWindows` seems to imply that this gate is needed to use `RuntimeClasses` for Windows which is not the case.

Reply:
Agree that `RuntimeClassesForWindows` can be misleading.
Also, should we try and make this enhancement more os-agnostic but commit to supporting windows checks? 
If so maybe `ConsiderOSDuringAdmission` or something more generic might make more sense?

Reply:
I don't have any issues with whatever name you folks come up with  :)

Reply:
`OSIdentificationDuringPodAdmission`?